graphqlapi/src/main/java/io/stargate/graphql/web/resources/GraphqlCache.java

```diff
@@ -80,33 +90,54 @@ public GraphQL getDdl() {
     return ddlGraphql;
   }
 
+  public GraphQL getSchemaFirstAdminGraphql() {
+    return schemaFirstAdminGraphql;
+  }
+
   public GraphQL getDml(
-      String keyspaceName, AuthenticationSubject subject, Map<String, String> headers) {
+      String keyspaceName, AuthenticationSubject subject, Map<String, String> headers)
+      throws Exception {
+
+    DataStore dataStore = buildUserDatastore(subject, headers);
+    SchemaSource latestSource = new SchemaSourceDao(dataStore).getLatestVersion(keyspaceName);
+
     String decoratedKeyspaceName = persistence.decorateKeyspaceName(keyspaceName, headers);
-    return dmlGraphqls.computeIfAbsent(
-        decoratedKeyspaceName, __ -> buildDmlGraphql(keyspaceName, subject, headers));
-  }
+    DmlGraphqlHolder currentHolder = dmlGraphqls.get(decoratedKeyspaceName);
+    if (currentHolder != null && currentHolder.isSameVersionAs(latestSource)) {
+      LOG.trace("Returning cached schema for {}", decoratedKeyspaceName);
+      return currentHolder.getGraphql();
+    }
 
-  public GraphQL getDefaultDml(Map<String, String> headers) {
-    return defaultKeyspace == null ? null : getDml(defaultKeyspace, null, headers);
+    DmlGraphqlHolder newHolder =
+        new DmlGraphqlHolder(latestSource, dataStore.schema().keyspace(keyspaceName));
+    boolean installed =
+        (currentHolder == null)
+            ? dmlGraphqls.putIfAbsent(decoratedKeyspaceName, newHolder) == null
+            : dmlGraphqls.replace(decoratedKeyspaceName, currentHolder, newHolder);
+
+    if (installed) {
+      LOG.debug("Installing new version for {} in the cache", decoratedKeyspaceName);
+      newHolder.init();
+      return newHolder.getGraphql();
+    }
+
+    LOG.debug(
+        "Got beat installing new version for {} in the cache, fetching again",
+        decoratedKeyspaceName);
+    currentHolder = dmlGraphqls.get(decoratedKeyspaceName);
+    return currentHolder == null ? null : currentHolder.getGraphql();
   }
 
   public String getDefaultKeyspaceName() {
     return defaultKeyspace;
   }
 
-  private GraphQL buildDmlGraphql(
-      String keyspaceName, AuthenticationSubject subject, Map<String, String> headers) {
+  private DataStore buildUserDatastore(AuthenticationSubject subject, Map<String, String> headers) {
     DataStoreOptions dataStoreOptions =
         DataStoreOptions.builder().putAllCustomProperties(headers).build();
-    DataStore dataStore =
-        subject == null
-            ? dataStoreFactory.createInternal(dataStoreOptions)
-            : dataStoreFactory.create(subject.asUser(), dataStoreOptions);
-    Keyspace keyspace = dataStore.schema().keyspace(keyspaceName);
-    return (keyspace == null)
-        ? null
-        : newGraphql(SchemaFactory.newDmlSchema(authorizationService, keyspace, dataStoreFactory));
+    return subject == null
```

Comment:
when the `subject == null` can be `null`?

Reply:
Right, that was when there was a `getDefaultDml`. I'll  remove this.